src/systems/media-frames.js

```diff
@@ -0,0 +1,295 @@
+import { MediaType } from "../utils/media-utils";
+import { TEXTURES_FLIP_Y } from "../loaders/HubsTextureLoader";
+import { applyPersistentSync } from "../utils/permissions-utils";
+
+// TODO better handling for 3d objects
+function scaleForAspectFit(containerSize, itemSize) {
+  return Math.min(containerSize.x / itemSize.x, containerSize.y / itemSize.y);
+}
+
+const isCapturableByType = {
+  [MediaType.ALL]: function(el) {
+    return !!(el && el.components["media-loader"]);
+  },
+  [MediaType.ALL_2D]: function(el) {
+    return !!(el && (el.components["media-image"] || el.components["media-video"] || el.components["media-pdf"]));
+  },
+  [MediaType.MODEL]: function(el) {
+    return !!(el && el.components["gltf-model-plus"]);
+  },
+  [MediaType.IMAGE]: function(el) {
+    return !!(el && el.components["media-image"]);
+  },
+  [MediaType.VIDEO]: function(el) {
+    return !!(el && el.components["media-video"]);
+  },
+  [MediaType.PDF]: function(el) {
+    return !!(el && el.components["media-pdf"]);
+  }
+};
+
+const EMPTY_COLOR = 0x6fc0fd;
+const HOVER_COLOR = 0x2f80ed;
+const FULL_COLOR = 0x808080;
+
+const components = [];
+export class MediaFramesSystem {
+  constructor(physicsSystem, interactionSystem) {
+    this.physicsSystem = physicsSystem;
+    this.interactionSystem = interactionSystem;
+  }
+
+  tick() {
+    for (let i = 0; i < components.length; i++) {
+      const frame = components[i];
+
+      const bodyUUID = frame.el.components["body-helper"].uuid;
+      // TODO would be nice to fix the timing so we don't need to check this
+      if (!this.physicsSystem.bodyInitialized(bodyUUID)) continue;
+
+      const holdingSomethingCapturable = this.interactionSystem.isHoldingAnything(
+        isCapturableByType[frame.data.mediaType]
+      );
+
+      const guideMesh = frame.el.getObject3D("guide");
+      guideMesh.visible = holdingSomethingCapturable;
+
+      if (frame.data.targetId === "empty") {
```

Comment:
I would like to clean up the branching logic here a bit, but would like to do it as a followup PR since its tricky to get right and I don't want to restart the testing process now.